lib/event_controller.rb

```diff
@@ -5,62 +5,84 @@
 
 require 'message_handler'
 require 'tiny_logger'
+require 'response'
 
 class EventController < Sinatra::Base
   attr_reader :message_handler
 
   def initialize
-    @message_handler = MessageHandler.new
+    @message_handler = MessageHandler.new(response: Response.new.setup)
   end
 
   post '/events' do
-    request_data = JSON.parse(request.body.read)
+    verify_token
+    verify_url
 
-    verify_token(request_data)
-    verify_url(request_data)
-
-    begin
-      Thread.new { handle_event(request_data) }
-    rescue StandardError => error
-      Logger.alert(error)
+    if okay_to_handle_event?
+      handle_event
     end
 
     status 200
   end
 
   private
 
-  def verify_token(data)
-    if invalid_token?(data['token'])
-      halt 403, "Invalid Slack verification token received: #{data['token']}"
+  def okay_to_handle_event?
+    event_is_callback? && event_is_message? && !message_from_bot?
+  end
+
+  def verify_token
+    if invalid_token?
+      halt 403, "Invalid Slack verification token received: #{token}"
+    end
+  end
+
+  def verify_url
+    body request_data['challenge'].to_s if url_verification?
+  end
+
+  def handle_event
+    Thread.new do
+      Logger.info("Receives Data #{event}")
+      @message_handler.handle(team_id, event)
     end
+  rescue StandardError => error
+    Logger.alert(error)
   end
 
-  def invalid_token?(token)
+  def invalid_token?
     !SLACK_CONFIG[:slack_verification_token] == token
   end
 
-  def verify_url(data)
-    body data['challenge'].to_s if data['type'] == 'url_verification'
+  def url_verification?
+    request_data['type'] == 'url_verification'
   end
 
-  def handle_event(data)
-    if data['type'] == 'event_callback'
-      team_id = data['team_id']
-      event_data = data['event']
+  def event_is_callback?
+    request_data['type'] == 'event_callback'
+  end
 
-      if message?(event_data) && !message_from_bot?(event_data)
-        Logger.info("Receives Data #{event_data}")
-        @message_handler.handle(team_id, event_data)
-      end
-    end
+  def event_is_message?
+    event['type'] == 'message'
+  end
+
+  def message_from_bot?
+    event['bot_id']
+  end
+
+  def token
+    request_data['token']
+  end
+
+  def team_id
+    request_data['team_id']
   end
 
-  def message?(event_data)
-    event_data['type'] == 'message'
+  def event
+    request_data['event']
   end
 
-  def message_from_bot?(event_data)
-    event_data['bot_id']
+  def request_data
+    JSON.parse(request.body.read)
```

Comment:
` post '/events' do` was receiving `request` as an argument of the block, I'm not quite sure `request_data` does receive from anywhere. Is that a Ruby understanding that I am missing?

Reply:
The `request` object comes from `Sinatra::Base`, so it's accessible anywhere in the controller.

https://github.com/sinatra/sinatra/blob/master/lib/sinatra/base.rb#L25

You're right, we should add some integration test for this. 👍